File: lite/core/src/calculateSplit.ts
import type {
  SplitSession,
  SplitResult,
  PersonResult,
  ItemShare,
  ChargeShare,
  ReceiptItem,
  Charge,
} from './types'

// ── Integer-cent arithmetic ───────────────────────────────────────────────────
//
// All distribution is done in integer cents to avoid floating-point drift.
// `toCents` converts a dollar amount to the nearest integer cent.
// Results are divided back to dollars only when assembling the final output.

function toCents(dollars: number): number {
  return Math.round(dollars * 100)
}

/**
 * Distributes `totalCents` among recipients proportionally to `weights` using
 * the largest-remainder method. Guarantees sum(result) === totalCents exactly.
 *
 * Works correctly for negative totals (discounts): floors toward −∞ and the
 * remainder brings totals back up by 1 cent for the highest-fractional slots.
 *
 * Falls back to equal distribution when weightSum is 0 (e.g. nobody has items).
 */
export function distributeInCents(totalCents: number, weights: number[]): number[] {
  const n = weights.length
  if (n === 0) return []
  if (n === 1) return [totalCents]

  const weightSum = weights.reduce((s, w) => s + w, 0)

  // Equal fallback
  if (weightSum === 0) {
    const base = Math.floor(totalCents / n)
    const rem = totalCents - base * n
    return weights.map((_, i) => base + (i < rem ? 1 : 0))
  }

  const exact = weights.map((w) => (totalCents * w) / weightSum)
  const floors = exact.map(Math.floor)
  const remainder = totalCents - floors.reduce((s, v) => s + v, 0)

  // Hand the remaining cents to the slots with the largest fractional parts
  const ranked = exact
    .map((v, i) => ({ i, frac: v - Math.floor(v) }))
    .sort((a, b) => b.frac - a.frac || a.i - b.i)   // stable: ties resolved by index

  const result = [...floors]
  for (let k = 0; k < Math.abs(remainder); k++) {
    result[ranked[k].i] += Math.sign(remainder) === 0 ? 0 : 1
  }

  return result
}

// ── Public entry point ────────────────────────────────────────────────────────

export function calculateSplit(session: SplitSession): SplitResult {
  const { receipt, people, assignments, splitMode } = session

  const assignedItemIds = new Set(
    assignments.filter((a) => a.personIds.length > 0).map((a) => a.itemId)
  )
  const unassignedItems: ReceiptItem[] = receipt.items.filter(
    (item) => !assignedItemIds.has(item.id)
  )

  if (people.length === 0) {
    return { personResults: [], unassignedItems, assignedTotal: 0, receiptTotal: receipt.total }
  }

  return splitMode === 'equal'
    ? equalSplit(session, unassignedItems)
    : itemizedSplit(session, unassignedItems)
}

// ── Equal split ───────────────────────────────────────────────────────────────

function equalSplit(session: SplitSession, unassignedItems: ReceiptItem[]): SplitResult {
  const { receipt, people } = session
  const n = people.length
  const evenWeights = Array<number>(n).fill(1)

  const subtotalCentsPerPerson = distributeInCents(toCents(receipt.subtotal), evenWeights)

  const chargeDistributionsCents = receipt.charges.map((charge) =>
    distributeInCents(toCents(charge.amount), evenWeights)
  )

  const personResults: PersonResult[] = people.map((person, i) => {
    const chargeShares: ChargeShare[] = receipt.charges.map((charge, ci) => ({
      charge,
      amount: chargeDistributionsCents[ci][i] / 100,
    }))
    const chargesTotalCents = chargeDistributionsCents.reduce((s, d) => s + d[i], 0)
    const subtotalCents = subtotalCentsPerPerson[i]

    return {
      person,
      itemShares: [],
      chargeShares,
      subtotal: subtotalCents / 100,
      chargesTotal: chargesTotalCents / 100,
      total: (subtotalCents + chargesTotalCents) / 100,
    }
  })

  return {
    personResults,
    unassignedItems,
    assignedTotal: receipt.total,
    receiptTotal: receipt.total,
  }
}

// ── Itemized split ────────────────────────────────────────────────────────────

function itemizedSplit(session: SplitSession, unassignedItems: ReceiptItem[]): SplitResult {
  const { receipt, people, assignments } = session

  // ── 1. Distribute each item's price among its assigned people ─────────────

  type ItemShareCents = { item: ReceiptItem; amountCents: number; outOf: number }
  type ChargeShareCents = { charge: Charge; amountCents: number }

  const itemSharesMap = new Map<string, ItemShareCents[]>(people.map((p) => [p.id, []]))
  const subtotalCentsMap = new Map<string, number>(people.map((p) => [p.id, 0]))

  for (const assignment of assignments) {
    if (assignment.personIds.length === 0) continue
    const item = receipt.items.find((i) => i.id === assignment.itemId)
    if (!item) continue

    const n = assignment.personIds.length
    const shares = distributeInCents(toCents(item.totalPrice), Array<number>(n).fill(1))

    assignment.personIds.forEach((personId, idx) => {
      itemSharesMap.get(personId)?.push({ item, amountCents: shares[idx], outOf: n })
      subtotalCentsMap.set(personId, (subtotalCentsMap.get(personId) ?? 0) + shares[idx])
    })
  }

  // ── 2. Distribute charges ─────────────────────────────────────────────────
  //
  // - proportional: weight = person's item subtotal in cents
  // - equal: weight = 1 for all
  // - none: skip entirely (e.g. rounding adjustments printed on the receipt)

  const subtotalWeights = people.map((p) => subtotalCentsMap.get(p.id) ?? 0)
  const evenWeights = Array<number>(people.length).fill(1)

  const chargeSharesMap = new Map<string, ChargeShareCents[]>(people.map((p) => [p.id, []]))

  for (const charge of receipt.charges) {
    if (charge.splitStrategy === 'none') continue

    const weights = charge.splitStrategy === 'equal' ? evenWeights : subtotalWeights
    const amounts = distributeInCents(toCents(charge.amount), weights)

    people.forEach((person, i) => {
      chargeSharesMap.get(person.id)?.push({ charge, amountCents: amounts[i] })
    })
  }

  // ── 3. Assemble PersonResult[] ────────────────────────────────────────────

  const personResults: PersonResult[] = people.map((person) => {
    const rawItems = itemSharesMap.get(person.id) ?? []
    const rawCharges = chargeSharesMap.get(person.id) ?? []

    const itemShares: ItemShare[] = rawItems.map((s) => ({
      item: s.item,
      amount: s.amountCents / 100,
      outOf: s.outOf,
    }))

    const chargeShares: ChargeShare[] = rawCharges.map((s) => ({
      charge: s.charge,
      amount: s.amountCents / 100,
    }))

    const subtotalCents = rawItems.reduce((s, is) => s + is.amountCents, 0)
    const chargesTotalCents = rawCharges.reduce((s, cs) => s + cs.amountCents, 0)

    return {
      person,
      itemShares,
      chargeShares,
      subtotal: subtotalCents / 100,
      chargesTotal: chargesTotalCents / 100,
      total: (subtotalCents + chargesTotalCents) / 100,
    }
  })

  // assignedTotal is the sum of all person totals (excludes unassigned items)
  const assignedTotalCents = personResults.reduce(
    (s, r) => s + toCents(r.total),
    0
  )

  return {
    personResults,
    unassignedItems,
    assignedTotal: assignedTotalCents / 100,
    receiptTotal: receipt.total,
  }
}

File: cloud/api/splitleh/calculate_split.py
"""Port of lite/core/src/calculateSplit.ts."""

from __future__ import annotations

from splitleh.money import to_cents
from splitleh.types import (
    ChargeShare,
    ItemShare,
    PersonResult,
    ReceiptItem,
    SplitResult,
    SplitSession,
)


def distribute_in_cents(total_cents: int, weights: list[int | float]) -> list[int]:
    n = len(weights)
    if n == 0:
        return []
    if n == 1:
        return [total_cents]

    weight_sum = sum(weights)
    if weight_sum == 0:
        base = total_cents // n
        rem = total_cents - base * n
        return [base + (1 if i < rem else 0) for i in range(n)]

    exact = [(total_cents * w) / weight_sum for w in weights]
    floors = [int(x // 1) for x in exact]
    remainder = total_cents - sum(floors)

    ranked = sorted(
        enumerate(exact),
        key=lambda pair: (-(pair[1] - int(pair[1] // 1)), pair[0]),
    )
    result = floors[:]
    sign = 1 if remainder >= 0 else -1
    for k in range(abs(remainder)):
        result[ranked[k][0]] += sign
    return result


def calculate_split(session: SplitSession) -> SplitResult:
    receipt = session.receipt
    people = session.people
    assignments = session.assignments
    split_mode = session.split_mode

    assigned_ids = {
        a.item_id for a in assignments if a.person_ids
    }
    unassigned = [i for i in receipt.items if i.id not in assigned_ids]

    if not people:
        return SplitResult(
            person_results=[],
            unassigned_items=unassigned,
            assigned_total=0.0,
            receipt_total=receipt.total,
        )

    if split_mode == "equal":
        return _equal_split(session, unassigned)
    return _itemized_split(session, unassigned)


def _equal_split(session: SplitSession, unassigned: list[ReceiptItem]) -> SplitResult:
    receipt = session.receipt
    people = session.people
    n = len(people)
    even = [1] * n

    subtotal_per = distribute_in_cents(to_cents(receipt.subtotal), even)
    charge_per_charge = [
        distribute_in_cents(to_cents(c.amount), even) for c in receipt.charges
    ]

    person_results: list[PersonResult] = []
    for i, person in enumerate(people):
        charge_shares = [
            ChargeShare(charge=c, amount=charge_per_charge[ci][i] / 100)
            for ci, c in enumerate(receipt.charges)
        ]
        charges_total_cents = sum(charge_per_charge[ci][i] for ci in range(len(receipt.charges)))
        subtotal_cents = subtotal_per[i]
        person_results.append(
            PersonResult(
                person=person,
                item_shares=[],
                charge_shares=charge_shares,
                subtotal=subtotal_cents / 100,
                charges_total=charges_total_cents / 100,
                total=(subtotal_cents + charges_total_cents) / 100,
            )
        )

    return SplitResult(
        person_results=person_results,
        unassigned_items=unassigned,
        assigned_total=receipt.total,
        receipt_total=receipt.total,
    )


def _itemized_split(session: SplitSession, unassigned: list[ReceiptItem]) -> SplitResult:
    receipt = session.receipt
    people = session.people
    assignments = session.assignments

    item_shares: dict[str, list[tuple[ReceiptItem, int, int]]] = {p.id: [] for p in people}
    subtotal_cents: dict[str, int] = {p.id: 0 for p in people}

    for assignment in assignments:
        if not assignment.person_ids:
            continue
        item = next((i for i in receipt.items if i.id == assignment.item_id), None)
        if item is None:
            continue
        n = len(assignment.person_ids)
        shares = distribute_in_cents(to_cents(item.total_price), [1] * n)
        for idx, person_id in enumerate(assignment.person_ids):
            item_shares[person_id].append((item, shares[idx], n))
            subtotal_cents[person_id] += shares[idx]

    subtotal_weights = [subtotal_cents[p.id] for p in people]
    even = [1] * len(people)
    charge_shares: dict[str, list[tuple[object, int]]] = {p.id: [] for p in people}

    for charge in receipt.charges:
        if charge.split_strategy == "none":
            continue
        weights = even if charge.split_strategy == "equal" else subtotal_weights
        amounts = distribute_in_cents(to_cents(charge.amount), weights)
        for i, person in enumerate(people):
            charge_shares[person.id].append((charge, amounts[i]))

    person_results: list[PersonResult] = []
    for person in people:
        raw_items = item_shares[person.id]
        raw_charges = charge_shares[person.id]
        item_share_objs = [
            ItemShare(item=it, amount=amt / 100, out_of=out_of)
            for it, amt, out_of in raw_items
        ]
        charge_share_objs = [
            ChargeShare(charge=c, amount=amt / 100) for c, amt in raw_charges
        ]
        sub_c = sum(amt for _, amt, _ in raw_items)
        chg_c = sum(amt for _, amt in raw_charges)
        person_results.append(
            PersonResult(
                person=person,
                item_shares=item_share_objs,
                charge_shares=charge_share_objs,
                subtotal=sub_c / 100,
                charges_total=chg_c / 100,
                total=(sub_c + chg_c) / 100,
            )
        )

    assigned_total_cents = sum(to_cents(r.total) for r in person_results)
    return SplitResult(
        person_results=person_results,
        unassigned_items=unassigned,
        assigned_total=assigned_total_cents / 100,
        receipt_total=receipt.total,
    )
